src/twisted/test/test_defer.py

```diff
@@ -1678,6 +1681,107 @@ def test_fromCoroutineRequiresCoroutine(self) -> None:
         for thing in thingsThatAreNotCoroutines:
             self.assertRaises(defer.NotACoroutineError, Deferred.fromCoroutine, thing)
 
+    @pyunit.skipIf(_PYPY, "GC works differently on PyPy.")
+    def test_canceller_circular_reference_callback(self) -> None:
+        """
+        A circular reference between a `Deferred` and its canceller
+        is broken when the deferred is called.
+        """
+
+        # Create a canceller and weak reference to track if its been freed.
+        canceller = DummyCanceller()
+        weakCanceller = weakref.ref(canceller)
+
+        # Create a Deferred with the above canceller, adding a reference to the
+        # deferred to the canceller to create the circular reference.
+        deferred: Deferred[Any] = Deferred(canceller)
+        canceller.deferred = deferred
+        weakDeferred = weakref.ref(deferred)
+
+        deferred.callback(None)
+
+        del deferred
```

Comment:
```suggestion
        # We manually remove the reference from the local
        # function scope, to avoid interfering with the test.
        del deferred
```